api/find.go

```diff
@@ -0,0 +1,97 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/puppetlabs/wash/activity"
+	"github.com/puppetlabs/wash/api/rql"
+	"github.com/puppetlabs/wash/api/rql/ast"
+	apitypes "github.com/puppetlabs/wash/api/types"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+// swagger:route GET /fs/find find listEntries
+//
+// Recursively descends the given path returning it and its children.
+//
+//     Consumes:
+//     - application/json
+//
+//     Produces:
+//     - application/json
+//
+//     Schemes: http
+//
+//     Responses:
+//       200: entryList
+//       400: errorResp
+//       404: errorResp
+//       500: errorResp
+var findHandler = handler{fn: func(w http.ResponseWriter, r *http.Request) *errorResponse {
+	ctx := r.Context()
+	entry, path, errResp := getEntryFromRequest(r)
+	if errResp != nil {
+		return errResp
+	}
+
+	if !plugin.ListAction().IsSupportedOn(entry) {
+		return unsupportedActionResponse(path, plugin.ListAction())
+	}
+
+	minDepth, hasMinDepth, errResp := getIntParam(r.URL, "mindepth")
+	if errResp != nil {
+		return errResp
+	}
+	maxDepth, hasMaxDepth, errResp := getIntParam(r.URL, "maxdepth")
+	if errResp != nil {
+		return errResp
+	}
+	fullMeta, errResp := getBoolParam(r.URL, "fullmeta")
+	if errResp != nil {
+		return errResp
+	}
+	var rawQuery interface{}
+	if err := json.NewDecoder(r.Body).Decode(&rawQuery); err != nil {
+		if err != io.EOF {
+			return badRequestResponse(fmt.Sprintf("could not decode the RQL query: %v", err))
+		}
+		rawQuery = true
+	}
+	query := ast.Query()
+	if err := query.Unmarshal(rawQuery); err != nil {
```

Comment:
That doesn't feel like the normal definition of Unmarshal. More like a deep type cast.

Reply:
If this were unmarshaling json, you could still do the previous test with a `json.RawMessage`.

Reply:
Not sure what you mean.

Reply:
Everything around ASTNode's `Unmarshal` seems kind of complicated. Why first unmarshal to an `interface{}` then pass that to a function to handle it rather than unmarshal directly in that function? If we need to do partial unmarshaling, that's what `json.RawMessage` is meant to handle.

Reply:
I like thinking of the input as Go types than some standardized thing (so map[string]interface{} instead of as a JSON object). That makes it easer to support YAML ASTs or some other structured data representation.

I'm open to changing it, just not sure what a better representation would look like.

Also, there isn't any partial unmarshalling going on here so don't think `json.RawMessage` is necessary.